### PR TITLE
HIVE-25192: Nullify summary for non-native tables

### PR DESCRIPTION
**Pull request by dengzhhu653**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 1]
```diff
@@ -2572,18 +2569,24 @@ public static ContentSummary getInputSummary(final Context ctx, MapWork work, Pa
 
       int numExecutors = getMaxExecutorsForInputListing(ctx.getConf(), pathNeedProcess.size());
       if (numExecutors > 1) {
```
Comment (dengzhhu653) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2571]: The default value of `numExecutors` is 1 unless `pathNeedProcess.size()` > 1 and `hive.exec.input.listing.max.threads` sets to the value bigger than 1. We do not need the global lock when `numExecutors = 1`,  and can be converted to a finer-grained lock. In our cluster, we found that many handler threads were blocked by the global lock when one of federation namespaces failover.

**Comment by dengzhhu653** [comment]
Hi @pvary, @belugabehr, @sahilTakiar cloud you take a look if have secs?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 2]
```diff
@@ -2680,9 +2683,9 @@ public void run() {
                   total += estimator.estimate(jobConf, scanOp, -1).getTotalLength();
                 }
                 recordSummary(path, new ContentSummary(total, -1, -1));
-              } else {
-                // todo: should nullify summary for non-native tables,
-                // not to be selected as a mapjoin target
+              } else if (handler == null) {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: This seems unrelated for me. Maybe a different PR?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 1 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2571]: I am not sure that the problem is with our threads, and not with threads calling this method

Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2571]: The problem is that the static global block `INPUT_SUMMARY_LOCK` will block other threads that calling this method.  The `INPUT_SUMMARY_LOCK`  is used for avoiding number of threads out of control, but this method does not use thread pool to get the summaries by default, so the lock can be converted to a finer-grained lock, which only takes care of the creation and execution of the thread pool.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 2 above)
Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: Thank you @pvary.  The pr originally is trying to solve the non-native table that participates in `join`, such table will be created with a path, but have no data under it, so the non-native table may be selected as a mapjoin target, which should cause oom problem when building the hashtable. The changes take place in one file, so I put them together.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 1 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2571]: Thanks for the explanation!
This seems reasonable!

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 2 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: What happened before this change?

Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: Could you please help explain?

Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: Last two months, in our cluster we found that join a kudu table(mr engine) cloud make the hiveserver2 oom.  [HMSHandler](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java#L2385-L2407) seems create a path for the non-native table, so `fs.getContentSummary(path)` returns the file length with 0(if data does not store under the table dir like kudu, hbase), the optimizer will [collect](https://github.com/apache/hive/blob/7b3ecf617a6d46f48a3b6f77e0339fd4ad95a420/ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/AbstractJoinTaskDispatcher.java#L118) the sizes of join tables before converting the join to mapjoin(by default), if the non-native table is able to be a stream table(like inner join) and the length of `his data` is 0,  the non-native will be tagged as small table and a local task will be started to build the hashtable. If the non-native table has millions of records, the jvm may not be able to hold all these records in memory.

Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: If the handler is not null (we have some specific handler which is not an `InputEstimator`), then we do not add the file sizes of data size, but we still have a `totalLength`, `totalFileCount`, `totalDirectoryCount` as 0?
Do I miss something?

Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: The summary will be recorded in current [Context](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L2709) and the change maps the non-native path to null as it ignores [recording summary](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L2700). The [Optimizer](https://github.com/apache/hive/blob/7b3ecf617a6d46f48a3b6f77e0339fd4ad95a420/ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/AbstractJoinTaskDispatcher.java#L133) retrieves the summary from the same Context instance, as it is null, indicates that the table is not a good choice for [mapjoin target](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L2757-L2760).

**Review by pvary — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 2 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: Thanks for the explanation and the patience! 😄

Comment (dengzhhu653) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:2686]: 😄 Thank you for the review!
